### PR TITLE
Fix option parsing

### PR DESCRIPTION
**Pull request by BobDotCom**
## Summary

Fixes option parsing... why was it like this in the first place

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [x] If `type: ignore` comments were used, a comment is also left explaining why
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

File: discord/commands/core.py [hunk 1]
```diff
@@ -758,30 +761,45 @@ async def _invoke(self, ctx: ApplicationContext) -> None:
             arg = arg["value"]
 
             # Checks if input_type is user, role or channel
-            if SlashCommandOptionType.user.value <= op.input_type.value <= SlashCommandOptionType.role.value:
-                if ctx.guild is None and op.input_type.name == "user":
-                    _data = ctx.interaction.data["resolved"]["users"][arg]
-                    _data["id"] = int(arg)
-                    arg = User(state=ctx.interaction._state, data=_data)
+            OT = SlashCommandOptionType
```
Comment (krittick) [inline discord/commands/core.py:764]:
```suggestion
```

File: discord/commands/core.py [hunk 2]
```diff
@@ -758,30 +761,45 @@ async def _invoke(self, ctx: ApplicationContext) -> None:
...
+            if op.input_type in (OT.user, OT.role, OT.channel, OT.attachment, OT.mentionable):
```
Comment (krittick) [inline discord/commands/core.py:765]:
```suggestion
            if op.input_type in (
                SlashCommandOptionType.user,
                SlashCommandOptionType.role,
                SlashCommandOptionType.channel,
                SlashCommandOptionType.attachment,
                SlashCommandOptionType.mentionable,
            ):
```

File: discord/commands/core.py [hunk 3]
```diff
@@ -758,30 +761,45 @@ async def _invoke(self, ctx: ApplicationContext) -> None:
...
+                resolved = ctx.interaction.data.get("resolved", {})
+                if (
+                        op.input_type in (OT.user, OT.mentionable)
```
Comment (krittick) [inline discord/commands/core.py:768]:
```suggestion
                        op.input_type in (SlashCommandOptionType.user, SlashCommandOptionType.mentionable)
```

File: discord/commands/core.py [hunk 4]
```diff
@@ -758,30 +761,45 @@ async def _invoke(self, ctx: ApplicationContext) -> None:
...
+                        and (_data := resolved.get("members", {}).get(arg)) is not None):
+                    # The option type is a user, we resolved a member from the snowflake and assigned it to _data
+                    if (_user_data := resolved.get("users", {}).get(arg)) is not None:
+                        # We resolved the user from the user id
+                        _data["user"] = _user_data
+                    arg = Member(state=ctx.interaction._state, data=_data, guild=ctx.guild)
+                elif op.input_type is OT.mentionable:
```
Comment (krittick) [inline discord/commands/core.py:775]:
```suggestion
                elif op.input_type is SlashCommandOptionType.mentionable:
```

File: discord/commands/core.py [hunk 5]
```diff
@@ -758,30 +761,45 @@ async def _invoke(self, ctx: ApplicationContext) -> None:
...
+                    if (_data := resolved.get(f"users", {}).get(arg)) is not None:
+                        arg = User(state=ctx.interaction._state, data=_data)
+                    elif (_data := resolved.get(f"roles", {}).get(arg)) is not None:
+                        arg = Role(state=ctx.interaction._state, data=_data, guild=ctx.guild)
+                    else:
+                        arg = Object(id=int(arg))
+                elif (_data := resolved.get(f"{op.input_type.name}s", {}).get(arg)) is not None:
+                    obj_type = None
+                    kw = {}
+                    if op.input_type is OT.user:
```
Comment (krittick) [inline discord/commands/core.py:785]:
```suggestion
                    if op.input_type is SlashCommandOptionType.user:
```

File: discord/commands/core.py [hunk 6]
```diff
@@ -758,30 +761,45 @@ async def _invoke(self, ctx: ApplicationContext) -> None:
...
+                        obj_type = User
+                    elif op.input_type is OT.role:
```
Comment (krittick) [inline discord/commands/core.py:787]:
```suggestion
                    elif op.input_type is SlashCommandOptionType.role:
```

File: discord/commands/core.py [hunk 7]
```diff
@@ -758,30 +761,45 @@ async def _invoke(self, ctx: ApplicationContext) -> None:
...
+                        obj_type = Role
+                        kw["guild"] = ctx.guild
+                    elif op.input_type is OT.channel:
```
Comment (krittick) [inline discord/commands/core.py:790]:
```suggestion
                    elif op.input_type is SlashCommandOptionType.channel:
```

File: discord/commands/core.py [hunk 8]
```diff
@@ -758,30 +761,45 @@ async def _invoke(self, ctx: ApplicationContext) -> None:
...
+                        obj_type = _guild_channel_factory(_data["type"])
+                        kw["guild"] = ctx.guild
+                    elif op.input_type is OT.attachment:
```
Comment (krittick) [inline discord/commands/core.py:793]:
```suggestion
                    elif op.input_type is SlashCommandOptionType.attachment:
```

**Review by krittick — DISMISSED** [review]
We should be clearer in using variable names, and this method still keeps the line length under 120.

**Comment by BobDotCom** [comment]
# Rationale
The API sends resolved users (and other data types such as channels and roles) in the interaction data. This PR fixes option parsing to use that resolved data instead of using `utils.get_or_fetch` on the IDs passed.
## Benchmarks
All times are relative times (i.e. they can be negative) but they are accurate when compared against each other. 
Above red line: Before
Below red line: After
All benchmarks are run with the member intent disabled.
### One user
In this benchmark we can see that the previous implementation slightly outperforms for the bots own user as it is always cached. The new implementation performs ~100-150ms faster for other users.

Screen Shot 2022-04-04 at 3 30 20 PM

### Five users
In this benchmark the bot user performs very similarly, but other users pull ahead by up to one second (1000ms).

Screen Shot 2022-04-04 at 3 44 09 PM

**Review by Dorukyum — APPROVED** [review]

**Review by Lulalaby — APPROVED** [review]

**Pull request merged**